library/kani_macros/src/sysroot/contracts.rs

```diff
@@ -237,30 +356,206 @@ fn rename_argument_occurences(sig: &syn::Signature, attr: &mut Expr) -> HashMap<
     arg_idents
 }
 
+/// The information needed to generate the bodies of check and replacement
+/// functions that integrate the conditions from this contract attribute.
+struct ContractConditionsHandler {
+    /// Information specific to the type of contract attribute we're expanding.
+    condition_type: ContractConditionsType,
+    /// The contents of the attribute.
+    attr: Expr,
+    /// Body of the function this attribute was found on.
+    body: Block,
+    /// An unparsed, unmodified copy of `attr`, used in the error messages.
+    attr_copy: TokenStream2,
+}
+
+/// Information needed for generating check and replace handlers for different
+/// contract attributes.
+enum ContractConditionsType {
+    Requires,
+    Ensures {
+        /// Translation map from original argument names to names of the copies
+        /// we will be emitting.
+        argument_names: HashMap<Ident, Ident>,
+    },
+}
+
+impl ContractConditionsType {
+    /// Constructs a [`Self::Ensures`] from the signature of the decorated
+    /// function and the contents of the decorating attribute.
+    ///
+    /// Renames the [`Ident`]s used in `attr` and stores the translation map in
+    /// `argument_names`.
+    fn new_ensures(sig: &Signature, attr: &mut Expr) -> Self {
+        let argument_names = rename_argument_occurrences(sig, attr);
+
+        ContractConditionsType::Ensures { argument_names }
+    }
+}
+
+impl ContractConditionsHandler {
+    /// Initialize the handler. Constructs the required
+    /// [`ContractConditionsType`] depending on `is_requires`.
+    fn new(
+        is_requires: bool,
+        mut attr: Expr,
+        fn_sig: &Signature,
+        fn_body: Block,
+        attr_copy: TokenStream2,
+    ) -> Self {
+        let condition_type = if is_requires {
+            ContractConditionsType::Requires
+        } else {
+            ContractConditionsType::new_ensures(fn_sig, &mut attr)
+        };
+
+        Self { condition_type, attr, body: fn_body, attr_copy }
+    }
+
+    /// Create the body of a check function.
+    ///
+    /// Wraps the conditions from this attribute around `self.body`.
+    fn make_check_body(&self, sig: &Signature) -> TokenStream2 {
+        let attr = &self.attr;
+        let attr_copy = &self.attr_copy;
+        let call_to_prior = &self.body;
+        let return_type = return_type_to_type(&sig.output);
+        match &self.condition_type {
+            ContractConditionsType::Requires => quote!(
+                kani::assume(#attr);
+                #call_to_prior
+            ),
+            ContractConditionsType::Ensures { argument_names } => {
+                let (arg_copies, copy_clean) = make_unsafe_argument_copies(&argument_names);
+                let attr = &self.attr;
```

Comment:
this is duplicated

Reply:
What do you mean?

Reply:
Line 419:
```rust
        let attr = &self.attr;
```

Reply:
I get that, but you already did that on line 419.

Reply:
Oh you're right.